listener/queue.py

```diff
@@ -0,0 +1,95 @@
+"""
+Listener Queue
+"""
+import asyncio
+
+from asyncpg.pool import Pool
+
+from common.mqueue import MQWriter
+from common.logging import get_logger
+from .common import InventoryMsgType, QueueItem, CFG, create_task_and_log
+from .processor import ListenerItemProcessor
+
+LOGGER = get_logger(__name__)
+
+
+class ListenerQueue:
+    """Listener async queue for message waiting"""
+
+    def __init__(self, db_pool: Pool, payload_tracker: MQWriter, loop: asyncio.BaseEventLoop):
+        """Constructor"""
+        self._queue = {}
+        self.item_processor = ListenerItemProcessor(db_pool, payload_tracker)
+        self.loop = loop
+
+    async def init(self):
+        """Async constructor"""
+        await self.item_processor.init()
+
+    async def stop(self):
+        """Stop queue"""
+        await self.item_processor.stop()
+
+    def push_inventory_msg(self, org_id: str, inventory_id: str, inventory_msg: dict, msg_type: InventoryMsgType):
+        """"Push listener message to queue"""
+        # on system delete attempt to delete all waiting messages from queue
+        # and also execute it immediately
+        if msg_type is InventoryMsgType.DELETE:
+            del self._queue[(org_id, inventory_id)]
+
+            item = QueueItem(inventory_msg, msg_type, None)
+            self.loop.create_task(self.item_processor.process_item(item))
+
+            return
+
+        # otherwise push to waiting queue
+        is_updated = False
+
+        item = self._queue.get((org_id, inventory_id), None)
+        if item is None:
+            item = QueueItem(inventory_msg, msg_type, None)
+        else:
+            is_updated = True
+            item.inventory_msg = inventory_msg
+            item.inventory_msg_type = msg_type
+
+        self._queue[(org_id, inventory_id)] = item
+        LOGGER.debug("Pushing listener msg to queue for system: %s, org_id: %s", inventory_id, org_id)
+
+        if item.advisor_msg is not None and item.inventory_msg is not None and item.inventory_msg_type is not InventoryMsgType.UNKNOWN:
```

Comment:
True, will change. `{}` for any of these messages cannot happen, because before calling these functions, the msgs gets validated for `org_id`, `display_name` fields, etc.